src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java

```diff
@@ -0,0 +1,247 @@
+
+    // Look up the directory entry by id. If it doesn't exist, we're done
+    TableEntity leafEntity = lookupByFileId(tableServiceClient, fileId);
+    if (leafEntity == null) {
+      return false;
+    }
+
+    List<TableTransactionAction> deleteList = new ArrayList<>();
+    TableTransactionAction t =
+        new TableTransactionAction(TableTransactionActionType.DELETE, leafEntity);
+    deleteList.add(t);
+
+    FireStoreDirectoryEntry leafEntry = FireStoreDirectoryEntry.fromTableEntity(leafEntity);
+    String datasetId = leafEntry.getDatasetId();
+    String lookupPath = fileMetadataUtils.makeLookupPath(leafEntry.getPath());
+    while (!lookupPath.isEmpty()) {
+      // Count the number of entries with this path as their directory path
+      // A value of 1 means that the directory will be empty after its child is
+      // deleted, so we should delete it also.
+      ListEntitiesOptions options =
+          new ListEntitiesOptions()
+              .setFilter(
+                  String.format(
+                      "path eq '%s'", fileMetadataUtils.makePathFromLookupPath(lookupPath)));
+      PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
+      int size = List.of(entities).size();
+      if (size > 1) {
+        break;
+      }
+      TableEntity entity =
+          lookupByFilePath(tableServiceClient, datasetId, encodePathAsAzureRowKey(lookupPath));
+      deleteList.add(new TableTransactionAction(TableTransactionActionType.DELETE, entity));
+      lookupPath = fileMetadataUtils.getDirectoryPath(lookupPath);
+    }
+    tableClient.submitTransaction(deleteList);
+    return true;
+  }
+
+  // Each dataset/snapshot has its own set of tables, so we delete the entire directory entry table
+  public void deleteDirectoryEntriesFromCollection(TableServiceClient tableServiceClient) {
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    tableClient.deleteTable();
+  }
+
+  // Returns null if not found - upper layers do any throwing
+  public FireStoreDirectoryEntry retrieveById(
+      TableServiceClient tableServiceClient, String fileId) {
+    TableEntity entity = lookupByFileId(tableServiceClient, fileId);
+    return Optional.ofNullable(entity)
+        .map(d -> FireStoreDirectoryEntry.fromTableEntity(entity))
+        .orElse(null);
+  }
+
+  // Returns null if not found - upper layers do any throwing
+  public FireStoreDirectoryEntry retrieveByPath(
+      TableServiceClient tableServiceClient, String datasetId, String fullPath) {
+    String lookupPath = fileMetadataUtils.makeLookupPath(fullPath);
+    TableEntity entity = lookupByFilePath(tableServiceClient, datasetId, lookupPath);
+    return Optional.ofNullable(entity)
+        .map(d -> FireStoreDirectoryEntry.fromTableEntity(entity))
+        .orElse(null);
+  }
+
+  public List<String> validateRefIds(
+      TableServiceClient tableServiceClient, List<String> refIdArray) {
```

Comment:
Should we make the fileId fields GUID (UUID)?  in this case this method would take `List<UUID>`